In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.model_selection import KFold,train_test_split,cross_val_score
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

In [2]:
data = pd.read_csv(r"diabetes.csv")

In [3]:
#EDA

In [4]:
data.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

In [5]:
#data.drop(['PassengerId','Name','Ticket','Cabin'],axis=1,inplace=True)

In [6]:
#data = data.ffill()

In [7]:
data.dtypes

Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object

In [8]:
#le = LabelEncoder()
#data[['Sex','Embarked']] = data[['Sex','Embarked']].apply(le.fit_transform)

In [16]:
x = data.drop('Outcome',axis=1)
X = StandardScaler().fit_transform(x)
y = data.Outcome

In [17]:
kf = KFold(n_splits=12)     #5 fold cross validation

In [18]:
s_model = SVC()

In [19]:
D_model = DecisionTreeClassifier()

In [20]:
R_model = RandomForestClassifier(n_estimators=500)

In [21]:
#5 fold validation of svm

In [22]:
li_s = []
for train_index,test_index in kf.split(x,y):
    xtrain,xtest,ytrain,ytest = x.loc[train_index],x.loc[test_index],y.loc[train_index],y.loc[test_index]
    s_model.fit(xtrain,ytrain)
    li_s.append(s_model.score(xtest,ytest))
    
print(li_s)
print("mean: ",sum(li_s)/len(li_s))


[0.6875, 0.78125, 0.75, 0.703125, 0.65625, 0.78125, 0.8125, 0.734375, 0.84375, 0.84375, 0.765625, 0.75]
mean:  0.7591145833333334


In [23]:
li_D = []
for train_index,test_index in kf.split(x,y):
    x_train,x_test,y_train,y_test = x.loc[train_index],x.loc[test_index],y.loc[train_index],y.loc[test_index]
    D_model.fit(x_train,y_train)
    li_D.append(D_model.score(x_test,y_test))
    
print(li_D)
print("mean : ",sum(li_D)/len(li_D))

[0.578125, 0.796875, 0.796875, 0.71875, 0.640625, 0.765625, 0.734375, 0.796875, 0.765625, 0.796875, 0.625, 0.796875]
mean :  0.734375


In [24]:
li_s = []
for train_index,test_index in kf.split(x,y):
    xtrain,xtest,ytrain,ytest = x.loc[train_index],x.loc[test_index],y.loc[train_index],y.loc[test_index]
    R_model.fit(xtrain,ytrain)
    li_s.append(R_model.score(xtest,ytest))
    
print(li_s)
print("mean: ",sum(li_s)/len(li_s))

[0.6875, 0.828125, 0.8125, 0.6875, 0.6875, 0.796875, 0.796875, 0.78125, 0.890625, 0.859375, 0.671875, 0.78125]
mean:  0.7734375


In [25]:
models = [s_model,D_model,R_model]
names = ['svm','DecisionTree','Randomforest']
for i in range(len(models)):
    li = []
    for train_index,test_index in kf.split(x,y):
        x_train,x_test,y_train,y_test = x.loc[train_index],x.loc[test_index],y.loc[train_index],y.loc[test_index]
        models[i].fit(x_train,y_train)
        li.append(models[i].score(x_test,y_test))
        
    print(li)
    #print("mean : ",names[i],sum(li)/len(li))
    print("mean : ",names[i],max(li))

[0.6875, 0.78125, 0.75, 0.703125, 0.65625, 0.78125, 0.8125, 0.734375, 0.84375, 0.84375, 0.765625, 0.75]
mean :  svm 0.84375
[0.609375, 0.765625, 0.734375, 0.65625, 0.65625, 0.6875, 0.703125, 0.78125, 0.75, 0.8125, 0.640625, 0.84375]
mean :  DecisionTree 0.84375
[0.703125, 0.828125, 0.78125, 0.671875, 0.6875, 0.796875, 0.796875, 0.796875, 0.890625, 0.859375, 0.671875, 0.765625]
mean :  Randomforest 0.890625


Random forest gives 89% accuracy for 12 fold cross validation.